Ast query with names and types with regex support

lukedirtwalker opened the pull request:
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39754947%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39754971%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39754997%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39755698%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756186%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756198%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756422%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756438%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756603%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756924%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39757064%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23issuecomment-141380300%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23issuecomment-141380300%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20fixed%20the%20things%20you%20mentioned%20except%20for%20the%20Regex%20stuff%2C%20which%20I%20noted%20for%20later%2C%20as%20I%27m%20not%20yet%20sure%20what%27s%20the%20best%20way%20for%20this.%22%2C%20%22created_at%22%3A%20%222015-09-18T08%3A27%3A26Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20a9f9ec26c1c459e43721533c671ce3158566a0d8%20InformationScripting/src/parsing/QueryBuilder.cpp%2029%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39754971%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22magic%202%22%2C%20%22created_at%22%3A%20%222015-09-17T14%3A55%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/QueryBuilder.cpp%3AL81-94%22%7D%2C%20%22Pull%20a9f9ec26c1c459e43721533c671ce3158566a0d8%20ModelBase/src/util/NameResolver.cpp%2014%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756603%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22A%20node%20can%20only%20ever%20belong%20to%20one%20manager.%20Thus%20if%20%60root%60%20is%20provided%2C%20you%20should%20not%20loop%20over%20all%20managers.%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A07%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20ModelBase/src/util/NameResolver.cpp%3AL40-47%22%7D%2C%20%22Pull%20a9f9ec26c1c459e43721533c671ce3158566a0d8%20InformationScripting/src/parsing/QueryBuilder.cpp%2025%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39754947%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22There%20are%20a%20lot%20of%20magical%20numbers%20here.%202%20should%20be%20inferred%20somehow%2C%20from%20the%20OPEN/CLOSE_SCOPE_SYMBOL%2C%20not%20a%20hardcoded%20constant.%22%2C%20%22created_at%22%3A%20%222015-09-17T14%3A55%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%20should%20be%20a%20class%20invariant%20that%20OPEN/CLOSE_SCOPE_SYMBOL%20have%20this%20length.%20Not%20sure%20how%20to%20implement%20this%20nicely.%5Cr%5Cn%5Cr%5CnFix%20a%20constant%20and%20iterate%20at%20construction%20over%20the%20symbols%20and%20check%20their%20length%3F%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A03%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22At%20the%20very%20least%2C%20define%20a%20static%20global%20constant%20at%20the%20beginning%20of%20this%20file%20%3D%202%20and%20use%20that%20throughout.%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A09%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/QueryBuilder.cpp%3AL81-94%22%7D%2C%20%22Pull%20a9f9ec26c1c459e43721533c671ce3158566a0d8%20InformationScripting/src/queries/AstQuery.cpp%20117%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756198%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20think%20at%20some%20point%20we%20should%20control%20the%20type%20of%20matcher%20based%20on%20command%20arguments.%20E.g.%20no%20argument%20-%3E%20simple%20substring%20matcher%20%28%2Atext%2A%29.%20Otherwise%20a%20fullblown%20regex%20matcher.%20But%20maybe%20some%20other%20smart%20option%20that%20detects%20the%20type%20of%20pattern%20is%20also%20fine.%20What%20do%20you%20think%3F%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A04%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Maybe%20some%20special%20prefix.%20How%20is%20this%20in%20bash%20%3F%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A05%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22not%20sure%2C%20but%20I%27ve%20definitely%20seen%20it%20for%20some%20commands%20%28e.g.%20%60sed%60%20uses%20%60-r%60%29%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A10%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL300-318%22%7D%2C%20%22Pull%20a9f9ec26c1c459e43721533c671ce3158566a0d8%20InformationScripting/src/queries/AstQuery.cpp%2093%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39755698%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Don%27t%20use%20the%20name%20%60result%60%20for%20something%20that%20is%20not%20the%20result%20of%20the%20method.%20It%20can%20be%20confusing.%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A00%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL242-282%22%7D%2C%20%22Pull%20a9f9ec26c1c459e43721533c671ce3158566a0d8%20InformationScripting/src/parsing/QueryBuilder.cpp%2042%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39754997%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Magic%202/4%22%2C%20%22created_at%22%3A%20%222015-09-17T14%3A55%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/QueryBuilder.cpp%3AL107-114%22%7D%2C%20%22Pull%20a9f9ec26c1c459e43721533c671ce3158566a0d8%20InformationScripting/src/visualization/DefaultVisualizer.cpp%2011%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/150%23discussion_r39756438%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Shouldn%27t%20we%20add%20an%20overlay%20to%20%2Aall%2A%20visualizations%20of%20an%20item%2C%20not%20just%20the%20first%20one%3F%22%2C%20%22created_at%22%3A%20%222015-09-17T15%3A05%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/visualization/DefaultVisualizer.cpp%3AL104-121%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [x] <a href='#crh-comment-Pull a9f9ec26c1c459e43721533c671ce3158566a0d8 InformationScripting/src/parsing/QueryBuilder.cpp 25'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#discussion_r39754947'>File: InformationScripting/src/parsing/QueryBuilder.cpp:L81-94</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> There are a lot of magical numbers here. 2 should be inferred somehow, from the OPEN/CLOSE_SCOPE_SYMBOL, not a hardcoded constant.
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> It should be a class invariant that OPEN/CLOSE_SCOPE_SYMBOL have this length. Not sure how to implement this nicely.
  Fix a constant and iterate at construction over the symbols and check their length?
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> At the very least, define a static global constant at the beginning of this file = 2 and use that throughout.
- [x] <a href='#crh-comment-Pull a9f9ec26c1c459e43721533c671ce3158566a0d8 InformationScripting/src/parsing/QueryBuilder.cpp 29'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#discussion_r39754971'>File: InformationScripting/src/parsing/QueryBuilder.cpp:L81-94</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> magic 2
- [x] <a href='#crh-comment-Pull a9f9ec26c1c459e43721533c671ce3158566a0d8 InformationScripting/src/parsing/QueryBuilder.cpp 42'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#discussion_r39754997'>File: InformationScripting/src/parsing/QueryBuilder.cpp:L107-114</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Magic 2/4
- [x] <a href='#crh-comment-Pull a9f9ec26c1c459e43721533c671ce3158566a0d8 InformationScripting/src/queries/AstQuery.cpp 93'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#discussion_r39755698'>File: InformationScripting/src/queries/AstQuery.cpp:L242-282</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Don't use the name `result` for something that is not the result of the method. It can be confusing.
- [x] <a href='#crh-comment-Pull a9f9ec26c1c459e43721533c671ce3158566a0d8 InformationScripting/src/queries/AstQuery.cpp 117'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#discussion_r39756198'>File: InformationScripting/src/queries/AstQuery.cpp:L300-318</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> I think at some point we should control the type of matcher based on command arguments. E.g. no argument -> simple substring matcher (_text_). Otherwise a fullblown regex matcher. But maybe some other smart option that detects the type of pattern is also fine. What do you think?
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Maybe some special prefix. How is this in bash ?
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> not sure, but I've definitely seen it for some commands (e.g. `sed` uses `-r`)
- [x] <a href='#crh-comment-Pull a9f9ec26c1c459e43721533c671ce3158566a0d8 InformationScripting/src/visualization/DefaultVisualizer.cpp 11'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#discussion_r39756438'>File: InformationScripting/src/visualization/DefaultVisualizer.cpp:L104-121</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Shouldn't we add an overlay to _all_ visualizations of an item, not just the first one?
- [x] <a href='#crh-comment-Pull a9f9ec26c1c459e43721533c671ce3158566a0d8 ModelBase/src/util/NameResolver.cpp 14'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#discussion_r39756603'>File: ModelBase/src/util/NameResolver.cpp:L40-47</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> A node can only ever belong to one manager. Thus if `root` is provided, you should not loop over all managers.
- [x] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/150#issuecomment-141380300'>General Comment</a></b>
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> I fixed the things you mentioned except for the Regex stuff, which I noted for later, as I'm not yet sure what's the best way for this.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/150?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/150?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/150'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

lukedirtwalker committed:
Add a start node in NameResolver::mostLikelyMatches

This makes it possible to search down from a certain node.

lukedirtwalker committed:
Enable the querying by name in AstQuery

lukedirtwalker committed:
Introduce different open and close symbols for Queries

Otherwise we can not detect if we have an open symbol or a closing one.

lukedirtwalker committed:
Use SymbolMatcher for types in Queries

This enables even more flexible queries.

lukedirtwalker committed:
Print a warning when trying to show nodes without visualization

dimitar-asenov commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/150'>Refresh</a>.<a href='#crh-update-none'></a>

dimitar-asenov commented on InformationScripting/src/parsing/QueryBuilder.cpp:
```diff
@@ -81,12 +81,13 @@ QPair<QStringList, QList<QChar>> QueryBuilder::split(const QString& text, const
 	QString currentString;
 	QVector<int> openScopes(3, 0);
-	for (int i = 1; i < text.size() - 1; ++i)
+	for (int i = 2; i < text.size() - 2; ++i)
```
There are a lot of magical numbers here. 2 should be inferred somehow, from the OPEN/CLOSE_SCOPE_SYMBOL, not a hardcoded constant.

lukedirtwalker replied:
It should be a class invariant that OPEN/CLOSE_SCOPE_SYMBOL have this length. Not sure how to implement this nicely.

Fix a constant and iterate at construction over the symbols and check their length?

dimitar-asenov replied:
At the very least, define a static global constant at the beginning of this file = 2 and use that throughout.

dimitar-asenov commented on InformationScripting/src/parsing/QueryBuilder.cpp:
```diff
@@ -81,12 +81,13 @@ QPair<QStringList, QList<QChar>> QueryBuilder::split(const QString& text, const
 	{
 		QChar currentChar = text.at(i);
-		int openIndex = OPEN_SCOPE_SYMBOL.indexOf(currentChar);
+		QString scopeSymbol = text.mid(i, 2);
```
magic 2

dimitar-asenov commented on InformationScripting/src/parsing/QueryBuilder.cpp:
```diff
@@ -106,7 +107,7 @@ QPair<QStringList, QList<QChar>> QueryBuilder::split(const QString& text, const
 {
 	Q_ASSERT(typeOf(text) == Type::Query);
-	QStringList data = text.mid(1, text.size()-2).split(" ", QString::SkipEmptyParts);
+	QStringList data = text.mid(2, text.size()-4).split(" ", QString::SkipEmptyParts);
```
Magic 2/4

dimitar-asenov commented on InformationScripting/src/queries/AstQuery.cpp:
```diff
@@ -235,11 +242,40 @@ TupleSet AstQuery::typeQuery(QList<TupleSet> input, QString type)
+	// If we have a type argument filter the results:
+	const QString type = argParser_.value(NODETYPE_ARGUMENT_NAMES[0]);
+	Model::SymbolMatcher matcher = matcherFor(type);
+	for (auto result : matchingNodes)
```
Don't use the name `result` for something that is not the result of the method. It can be confusing.

dimitar-asenov commented on InformationScripting/src/queries/AstQuery.cpp:
```diff
@@ -264,12 +300,18 @@ void AstQuery::addCallInformation(TupleSet& ts, OOModel::Method* method, QList<O
 		ts.add({{"ast", node}});
 }
 
+Model::SymbolMatcher AstQuery::matcherFor(const QString& text)
```
I think at some point we should control the type of matcher based on command arguments. E.g. no argument -> simple substring matcher (_text_). Otherwise a fullblown regex matcher. But maybe some other smart option that detects the type of pattern is also fine. What do you think?

lukedirtwalker replied:
Maybe some special prefix. How is this in bash ?

dimitar-asenov replied:
not sure, but I've definitely seen it for some commands (e.g. `sed` uses `-r`)

lukedirtwalker replied:
I noted this for later.

dimitar-asenov commented on InformationScripting/src/visualization/DefaultVisualizer.cpp:
```diff
@@ -104,12 +104,17 @@ void DefaultVisualizer::visualize(const TupleSet& ts)
-			auto overlay = new Visualization::SelectionOverlay(
-						item, Visualization::SelectionOverlay::itemStyles().get(it.value() + "Highlight"));
-			item->addOverlay(overlay, HIGHLIGHT_OVERLAY_GROUP);
+			if (nodeVisualization != Visualization::Item::nodeItemsMap().end())
```
Shouldn't we add an overlay to _all_ visualizations of an item, not just the first one?

dimitar-asenov commented on ModelBase/src/util/NameResolver.cpp:
```diff
@@ -40,7 +40,7 @@ QList<QPair<QString, Node*>> NameResolver::mostLikelyMatches(const QString& node
 
 	for (auto manager : AllTreeManagers::instance().loadedManagers())
-		matches.append(findAllMatches(matcher, "", manager->root()));
+		matches.append(findAllMatches(matcher, "", root ? root : manager->root()));
```
A node can only ever belong to one manager. Thus if `root` is provided, you should not loop over all managers.

lukedirtwalker committed:
Refactor: Replace magic numbers in QueryBuilder with a const variable.

Also make OPEN/CLOSE_SCOPE_SYMBOLS lists a static class member, so it is
not free standing in the namespace.

lukedirtwalker committed:
Refactor: don't use the name result for a local variable
which is not returned

lukedirtwalker committed:
Show all visualizations for a node in DefaultVisualizer

lukedirtwalker committed:
Only loop over managers if we have no root argument in NameResolver

It doesn't make sense in the case we have a root since it stays the
same.

lukedirtwalker commented:
I fixed the things you mentioned except for the Regex stuff, which I noted for later, as I'm not yet sure what's the best way for this.
